### PR TITLE
add math-class group chat test

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
## Why are these changes needed?
this test is imported from #102 

After #294 it's possible to test the group chat result in a more robust way. 

By using `function_call` and adding a special prefix to messages, we can examine the conversation flow inside a group chat.

This test also provides an example for #133, which terminates group chat/two agent chat in a more controlable way

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

File: test/agentchat/test_groupchat.py [hunk 1]
```diff
@@ -49,6 +51,181 @@ def test_func_call_groupchat():
...
+    where teacher creates math questions and student resolves the questions.
+    teacher will create a question, student will resolve the question and tell teacher the answer.
+    If the answer is correct, teacher will create another question, otherwise, teacher will ask student to resolve the question again.
+    The class will end when teacher has created 3 questions.
+
+    This test case is created to test the following features:
+    - speaker selection should work under a continuous q&a scenario among two agents and GPT 3.5 model.
+    - admin should end the class when teacher has created 3 questions.
+    """
+    skip_if_openai_not_available()
```
Comment (sonichi) [inline test/agentchat/test_groupchat.py:66]: Nice. This is better than my old way. @rickyloynd-microsoft @thinkall @kevin666aa FYI.

**Review by sonichi — COMMENTED** [review]
Looks good structure wise.

**Comment by sonichi** [comment]
If you know which user can benefit from this PR, could you ask them to review it?

**Comment by LittleLittleCloud** [comment]
@sonichi. It's @afourney. I saw you just sent him a request

File: test/agentchat/test_groupchat.py [hunk 2]
```diff
@@ -49,6 +51,181 @@ def test_func_call_groupchat():
...
+                            "description": "terminate group chat message",
+                        },
+                    },
+                    "required": ["message"],
+                },
+            }
+        ],
+    }
+
+    def terminate_group_chat(message):
```
Comment (LittleLittleCloud) [inline test/agentchat/test_groupchat.py:102]: @afourney This is how to achieve a more robust terminating strategy via function_call, could you review it?

**Comment by sonichi** [comment]
@LittleLittleCloud I tried to add more reviewers who might be interested. If you know of any others, please invite them.

**Comment by miiiz** [comment]
hello. I try the test_group_chat_math_class() example but still can't make the group chat terminate. Could you please help take a look. I only change question number from 3 to 2.


--------------------------------------------------------------------------------
teacher (to chat_manager):

***** Suggested function Call: create_math_question *****
Arguments: 
{
  "question": "What is 2 + 3?",
  "i": 1
}
*********************************************************

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING FUNCTION create_math_question...
teacher (to chat_manager):

***** Response from calling function "create_math_question" *****
[QUESTION] this is question #1: What is 2 + 3?
*****************************************************************

--------------------------------------------------------------------------------
student (to chat_manager):

Teacher, the question is "What is 2 + 3?"

--------------------------------------------------------------------------------
student (to chat_manager):

***** Suggested function Call: answer_math_question *****
Arguments: 
{
  "answer": "5"
}
*********************************************************

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING FUNCTION answer_math_question...
student (to chat_manager):

***** Response from calling function "answer_math_question" *****
[ANSWER] 5
*****************************************************************

--------------------------------------------------------------------------------
teacher (to chat_manager):

Great job, student! The answer to "What is 2 + 3?" is indeed 5. Now, let's move on to the next question.
***** Suggested function Call: create_math_question *****
Arguments: 
{
  "question": "What is 4 - 2?",
  "i": 2
}
*********************************************************

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING FUNCTION create_math_question...
teacher (to chat_manager):

***** Response from calling function "create_math_question" *****
[QUESTION] this is question #2: What is 4 - 2?
*****************************************************************

--------------------------------------------------------------------------------
student (to chat_manager):

Teacher, the question is "What is 4 - 2?"

--------------------------------------------------------------------------------
student (to chat_manager):

***** Suggested function Call: answer_math_question *****
Arguments: 
{
  "answer": "2"
}
*********************************************************

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING FUNCTION answer_math_question...
student (to chat_manager):

***** Response from calling function "answer_math_question" *****
[ANSWER] 2
*****************************************************************

--------------------------------------------------------------------------------
Admin (to chat_manager):

***** Suggested function Call: terminate_group_chat *****
Arguments: 
{
  "message": "Great job, everyone! The group chat is now complete."
}
*********************************************************

--------------------------------------------------------------------------------

>>>>>>>> EXECUTING FUNCTION terminate_group_chat...
Admin (to chat_manager):

***** Response from calling function "terminate_group_chat" *****
[GROUPCHAT_TERMINATE] Great job, everyone! The group chat is now complete.
*****************************************************************

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
teacher (to chat_manager):

Great job, student! The answer to "What is 4 - 2?" is indeed 2. You have successfully answered both math questions correctly. Well done!

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
teacher (to chat_manager):

Great job, student! You have successfully answered both math questions correctly. Well done!

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
Admin (to chat_manager):

[COMPLETE]

--------------------------------------------------------------------------------
teacher (to chat_manager):

Great job, student! You have successfully answered both math questions correctly. Well done!

--------------------------------------------------------------------------------

**Comment by LittleLittleCloud** [comment]
@miiiz 
If you just copy the code example and run it on the main branch, the group chat won't exit even if `terminate_group_chat` function gets called.

This is because this PR also contains the following change which exits group chat if a message contains specific content.
image

**Comment by miiiz** [comment]
@LittleLittleCloud 
I actually run in this branch:[LittleLittleCloud:u/xiaoyun/addTest].
I checked that the GroupChatManager also has 'is_termination_msg' param. 


image

**Comment by miiiz** [comment]
@LittleLittleCloud I make these changes and works fine now.

type(reply) == dict
type(reply) == str

**Comment by LittleLittleCloud** [comment]
@afourney Can you review this PR when you have time

File: test/agentchat/test_groupchat.py (hunk 2 above)
Comment (gagb) [reply test/agentchat/test_groupchat.py:102]: @LittleLittleCloud I am having trouble understand why this strategy is more robust?

This strategy: one of the agents calls a group chat termination function. user proxy executes. manager detects termination string

Old strategy: one of the agent generates a termination string. manager detects.

**Comment by gagb** [comment]
Once we approve this PR I think we can make progress on #525 and #517. (cc @victordibia @pcdeadeasy)

File: test/agentchat/test_groupchat.py (hunk 2 above)
Comment (LittleLittleCloud) [reply test/agentchat/test_groupchat.py:102]: > If an agent is smart enough for suggested strategy then it should be able to also do old strategy?

Yes! Unfortunately, that's not the case in the real world, especially for `gpt-3.5-turbo`. Even for `gpt-4` agents, they might also fail to give the correct termination world when the conversation grows.

The new strategy (using termination_function_call) can make sure the group chat terminate correctly when that termination function_call get triggered. That strategy also works well on gpt-3.5-turbo which fined-tuned for function_call.

Comment (gagb) [reply test/agentchat/test_groupchat.py:102]: Got it -- the fact that function calls are prioritized makes it more robust.

**Review by gagb — APPROVED** [review]

File: autogen/agentchat/groupchat.py [hunk 3]
```diff
@@ -178,6 +180,15 @@ def run_chat(
...
                 break
+
+            if (
+                type(reply) == dict
+                and self._is_termination_msg(reply)
+                or type(reply) == str
+                and self._is_termination_msg({"content": reply})
+            ):
+                break
+
```
Comment (gagb) [inline autogen/agentchat/groupchat.py:191]: @LittleLittleCloud, should these lines be after `speaker.send` ??

Comment (sonichi) [reply autogen/agentchat/groupchat.py:191]: also, I usually use `isinstance()` instead of `type()`

Comment (gagb) [reply autogen/agentchat/groupchat.py:191]: > @LittleLittleCloud, should these lines be after `speaker.send` below??

Did this suggestion make sense?

Comment (LittleLittleCloud) [reply autogen/agentchat/groupchat.py:191]: nice catch

**Comment by ekzhu** [comment]
@LittleLittleCloud  bump so this PR doesn't get abandoned :D

**Pull request closed without merging**